server: Reuse the raw socket

Natolumin opened the pull request:
Creating, then destroying a raw socket for every packet is very heavy. We can open the socket and keep it when first opening the (udp) listening socket, so that we are able to reuse the raw socket for further packets

Instead of managing the packet socket with manual syscalls, we can use the https://github.com/mdlayher/raw library (which we already imported before anyway) which will give us portability for free.
This does however mean that we cannot support dynamically setting the interface differently for each packet (at the moment) as the library requires an interface when binding to the socket; so this requires binding to a specific interface at startup (rather than relying on out-of-band data when receiving packets to choose the outgoing interface).
For DHCP servers this is a very common behavior (and we even required it until not that long ago) so I'm happy to take that tradeoff until we can update the upstream library to support the other usecase if needed.

Finally, if we can't setup the raw socket, we can then fallback to sending over broadcast (which is allowed by the RFC), ensuring that the server can keep working without for example CAP_NET_RAW

Natolumin committed:
server: Reuse the raw socket

Creating, then destroying a raw socket for every packet is very heavy.
We can open the socket and keep it when first opening the listening
socket, so that we are able to reuse it

This however requires knowing and binding to an interface at socket
creation time (which is fairly standard for dhcpv4) due to a limitation
of the library we use

Signed-off-by: Anatole Denis <anatole@unverle.fr>

Natolumin committed:
server: Select an appropriate IP address for L2 unicast

When deferring to the system routing, the ip address is automatically
selected by the kernel.
We were previously using the message's ServerIPAddr, but that is
incorrect as a source address since that is the address of the next
server, and not necessarily the address of the current server or an
address that is routable by the client

Signed-off-by: Anatole Denis <anatole@unverle.fr>

insomniacslk commented:
> For DHCP servers this is a very common behavior (and we even required it until not that long ago) so I'm happy to take that tradeoff until we can update the upstream library to support the other usecase if needed.

I am OK with this behaviour

Natolumin committed:
server: Add logging for L2 capabilities

Signed-off-by: Anatole Denis <anatole@unverle.fr>

Natolumin committed:
server: Add stub sendEthernet for non-linux environments

With the deprecation of mdhlayer/raw and its replacement with
mdhlayer/packet, we lost BSD support for AF_PACKET sockets
Since the server code degrades gracefully if tryOpenRawSock fails,
compile a stub to let the server compile and start outside linux

Signed-off-by: Anatole Denis <anatole@unverle.fr>
